### ISSUE TITLE
Support duplicate lock acquisition

### ISSUE THREAD

**Issue by armon** [hashicorp/consul#876]
It should be a no-op to re-acquire a lock if its already held by the same session.

See: https://groups.google.com/forum/#!topic/consul-tool/Cuhr_DVXqZE

**Comment by nathan7** [comment]
I made myself a works-for-me implementation: https://github.com/nathan7/consul/commit/f182f04c8aa29705dc385b23722fbf18c6152abc
Along the way, I discovered that the serialisation mechanism doesn't seem to care for the difference between `[]byte{}` and `[]byte(nil)` — resulting in `"Value": null` in the HTTP API if you sent an empty PUT. This doesn't seem to be documented behaviour.
I mainly didn't want to add a new field to `structs.DirEnt`, given that it seems to be quite widely used.
I'd like to get this implemented and merged — I've run into this case a bunch: acquire a lock on some state, mutate it while you have the lock, having the lock and the state clearly be one and the same key.
(being able to pass the LockIndex on a PUT would be really neat — it makes it trivial to use Consul for a piece of single-reader/writer mutable state safely)
